pydoctor/test/test_zopeinterface.py

```diff
@@ -360,12 +366,64 @@ class Implementation:
     mod = fromText(src, modname='mod', systemcls=ZopeInterfaceSystem)
     C = mod.contents['C']
     impl = mod.contents['Implementation']
-    assert C.isinterface
-    assert C.kind == "Interface"
-    assert C.implementedby_directly == [impl]
+    assert not C.isinterface
+    assert C.kind == "Class"
     assert impl.implements_directly == ['mod.C']
     captured = capsys.readouterr().out
-    assert captured == "mod:5: probable interface mod.C not marked as such\n"
+    assert captured == 'mod:5: Class "mod.C" is not an interface\n'
+
+def test_implementer_not_found(capsys: CapSys) -> None:
+    """
+    An unknown class passed to @implementer is warned about if its full name
+    is part of our system.
+    """
+    src = '''
+    from zope.interface import implementer
+    from twisted.logger import ILogObserver
+    @implementer(ILogObserver, mod.INoSuchInterface)
+    class Implementation:
+        pass
+    '''
+    fromText(src, modname='mod', systemcls=ZopeInterfaceSystem)
+    captured = capsys.readouterr().out
+    assert captured == 'mod:4: Interface "mod.INoSuchInterface" not found\n'
+
+def test_implementer_reparented() -> None:
+    """
+    A class passed to @implementer can be found even when it is moved
+    to a different module.
+    """
+
+    system = ZopeInterfaceSystem()
+
+    mod_iface = fromText('''
+    from zope.interface import Interface
+    class IMyInterface(Interface):
+        pass
+    ''', modname='_private', system=system)
+
+    mod_export = fromText('', modname='public', system=system)
+
+    mod_impl = fromText('''
+    from zope.interface import implementer
+    from _private import IMyInterface
+    @implementer(IMyInterface)
+    class Implementation:
+        pass
+    ''', modname='app', system=system)
+
+    iface = mod_iface.contents['IMyInterface']
+    iface.reparent(mod_export, 'IMyInterface')
+    assert iface.fullName() == 'public.IMyInterface'
+    assert 'IMyInterface' not in mod_iface.contents
+
+    impl = mod_impl.contents['Implementation']
+    assert impl.implements_directly == ['_private.IMyInterface']
+    assert iface.implementedby_directly == []
+
+    system.postProcess()
+    assert impl.implements_directly == ['_private.IMyInterface']
```

Comment:
Should the post processing update the name of the interface to the reparented version? I picked "no" when writing this test, such that the code model reflects the source code. However, I'm not 100% sure this leads to the best possible output. I'll have to check which classes in Twisted are affected by this and how their output looks.

Reply:
An example is [`twisted.logger.FileLogObserver`](https://twistedmatrix.com/documents/current/api/twisted.logger.FileLogObserver.html). The relevant part of the code looks like this:

```py
from ._interfaces import ILogObserver

@implementer(ILogObserver)
class FileLogObserver:
    """
    Log observer that writes to a file-like object.
    """
```

So it imports the interface from its private module, but both the interface and `FileLogObserver` itself get re-exported to a public module.

In this particular case it's clear that the fact that `FileLogObserver` implements `ILogObserver` should be public. Should we assume that if a public interface is implemented, that should always be considered a public relationship? Or are there situations in which it makes sense to implement a public interface privately?

In the latter case, what would be a good way for the code to signal that? Importing the interface from its private module could be considered as a hint, but that would cause problems in cases like this, where importing from the public module would make imports very tricky. Since both the interface and the implementation are re-exported from the same module (`twisted/logger/__init__.py`), trying to import the interface from its public location would create a nasty import cycle where re-ordering imports could break the modules altogether.

So I think that for practical concerns alone, we should display all implemented public interfaces, regardless of whether they were imported from their public or private module. But I'm still interested in hearing whether this is also the right thing to do conceptually or whether we might want to support private implementation of public interfaces in the future.

Reply:
I pushed some changes that will update the names of reparented interfaces during post-processing, such that the generated API docs will always use the new (public) name.

Reply:
> So I think that for practical concerns alone, we should display all implemented public interfaces, regardless of whether they were imported from their public or private module.

+1

>  But I'm still interested in hearing whether this is also the right thing to do conceptually or whether we might want to support private implementation of public interfaces in the future.

I can't think of such an example.
I guess that is best to leave it as it is (show all interfaces with private name)

If someone needs a different behavior, they can create a ticket and we can discuss the exact usage.

---------

Note that we also have https://docs.twistedmatrix.com/en/latest/api/twisted.logger.FileLogObserver.html which is based on "trunk" 
I don't see the "Implements interfaces:" on that page